### PR TITLE
Add option for separable kernel in VRT format

### PR DESCRIPTION
**Pull request by mholling**
This pull requests adds the option to use a separable kernel for a KernelFilteredSource element in a VRT file.

A Gaussian blur kernel is the obvious case of a separable kernel. Instead of running the 2D Gaussian kernel over the image, you can run a 1D Gaussian kernel twice, once in the horizontal direction and once in the vertical direction. A separable kernel has O(r) runtime, instead of O(r^2), where r is the radius of the kernel. The speedup is very welcome for large-radius blurs.

To implement separable kernel filtering, I did not need to change the XML schema for the VRT format. Instead, a 1D separable kernel is automatically detected when the number of entries in the kernel list is the same as the specified 'Size' element (instead of Size^2 for the normal case).

If a separable kernel is present, the filtering loop is run twice, with the axes inverted for the second pass. Both the separable- and 2D-kernel cases share the same looping code, with the 2D case only looping once instead of twice. I refactored the core filtering loop to allow for this, using stride values to step across the buffer.

I changed the setup code in VRTFilteredSource::RasterIO to pad the output buffer as well as the source buffer. This is necessary to allow for the two-pass filtering.

I hope this enhancement is welcome! The speedups for blurring filters are really welcome. I was using a Gaussian filter to blur a DEM and found it to be quite slow. I expect blurring is a common use of kernel filtered VRT files. Other operations (e.g. sharpening kernels) would also benefit.

### CODE REVIEW

**Comment by mholling** [comment]
p.s. the first three commits are from a contribution I made last year; not sure why they're included in this pull request, they were already accepted.

File: gdal/frmts/vrt/vrt_tutorial.dox [hunk 1]
```diff
@@ -446,6 +446,23 @@ and the Coefs must have Size * Size entries separated by spaces.
       </Kernel>
     </KernelFilteredSource>
 \endcode
+
+Starting with GDAL 2.2, a separable kernel may also be used.  In this case the
```
Comment (rouault) [inline gdal/frmts/vrt/vrt_tutorial.dox]: 2.3 now

File: gdal/frmts/vrt/vrtfilters.cpp [hunk 2]
```diff
@@ -495,65 +489,69 @@ CPLErr VRTKernelFilteredSource::FilterData( int nXSize, int nYSize,
...
     {
+        float *pafSrcData = reinterpret_cast<float *>( pabySrcData );
+        float *pafDstData = reinterpret_cast<float *>( pabyDstData );
+
         int bHasNoData = FALSE;
         const float fNoData =
             static_cast<float>( m_poRasterBand->GetNoDataValue(&bHasNoData) );
 
-        for( int iY = 0; iY < nYSize; iY++ )
+        for( int nAxis = 0; m_bSeparable ? nAxis < 2 : nAxis < 1; ++nAxis)
```
Comment (rouault) [inline gdal/frmts/vrt/vrtfilters.cpp]: Perhaps introduce const int nAxisCount = m_bSeparable ? 2 : 1 and use it in the for condition

**Comment by mholling** [comment]
Updated as per your suggestions.

I started writing a test case, but I couldn't get the python bindings to build. Unfamiliar with python! I will have another go but it might take a while.

**Comment by rouault** [comment]
> I couldn't get the python bindings to build

```
sudo apt-get install python-devel (or similar to install python development packages)
cd swig/python
make
export PYTHONPATH=$PWD/swig/python/build/lib.linux-x86_64-2.7 (to adapt given your platform and python version)
```

**Comment by mholling** [comment]
No luck with running the tests; I got past the build error by deactivating my normal GDAL installation, but tests won't run ('ImportError: No module named _gdal'). Urgh.

I've added a test case anyway, cautiously optimistic that it'll pass.

**Comment by rouault** [comment]
> I got past the build error by deactivating my normal GDAL installation, but tests won't run. (Not finding libgdal.20.dylib. Urgh.)

export DY_LD_LIBRARY_PATH=/path/to/libgdal.20.dylib ?

**Comment by mholling** [comment]
OK, that did it. (It was DYLD_LIBRARY_PATH for me.) And the test passes. Thanks for your guidance. Let me know if there are any other issues.

**Comment by rouault** [comment]
Committed into r38200. Thanks

**Comment by mholling** [comment]
Great, thanks. GDAL's a great project, I'm please to be able to make a small contribution.

**Pull request closed without merging**